wcfsetup/install/files/lib/data/I18nDatabaseObjectList.class.php

```diff
@@ -0,0 +1,72 @@
+<?php
+namespace wcf\data;
+use wcf\system\exception\SystemException;
+use wcf\system\WCF;
+use wcf\util\StringUtil;
+
+/**
+ * Abstract class for a list of database objects with better sorting of i18n-based columns.
+ *
+ * @author	Florian Gail
+ * @copyright	2001-2017 WoltLab GmbH
+ * @license	GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package	WoltLabSuite\Core\Data
+ * @since	3.1
+ */
+abstract class I18nDatabaseObjectList extends DatabaseObjectList {
+	/**
+	 * Array of column names, that are eventually filled with language items.
+	 * Those additional joins are going to slow down the system; you should use this
+	 * class only when it's really needed.
+	 * The key represents the original field name.
+	 * The value represents the new field name containing the localized field-value.
+	 *
+	 * @example   [ 'title' => 'titleSortField' ]
+	 *
+	 * @var string[]
+	 */
+	public $i18nFields = [];
+	
+	/**
+	 * ID of the language that should be used
+	 * the value defaults to the current user's language
+	 *
+	 * @var integer
+	 */
+	public $languageID;
```

Comment:
This does not allow you to change the languageID, because it already is embedded into the string once the constructor finished. I suggest adding a `$languageID` parameter to the constructor (and possibly a `setSortingLanguage` – I'm open for a better method name – method that allows you to change the language after construction, obviously use that method in the constructor instead of repeating code).

Reply:
How do you want to change the ID after construction? It's an integer within a string that could contain anything. That would be possible if you handle the custom joins/selects and those for i18n in different strings/arrays and rebuild the query right before every execution like this:
```php
$tmpSelects = $this->sqlSelects; // the same for joins…
$this->sqlSelects .= (!empty($this->sqlSelects) ? ', ' : '') . $this->i18nSelects;
// query
$this->sqlSelects = $tmpSelects;
```